### PR TITLE
Refactor LinearActQuantizedTensor

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
* rename to LinearActivationQuantizedTensor
* using `implements` util to implement torch function and torch dispatch overwrites
* refactored tensor subclass dispatch related utility functions: `_implements` and added two dispatch helpers: `_dispatch__torch_function__` and `_dispatch__torch_dispatch__` that saves more lines of code from users

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: tutorials/calibration_flow/static_quant.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 from torchao.dtypes import to_affine_quantized_static
 from torchao.quantization.utils import compute_error
 from torchao.quantization import quantize_
-from torchao.quantization.subclass import to_linear_act_quantized
+from torchao.quantization.subclass import to_linear_activation_quantized
```
Comment (jcaip) [inline tutorials/calibration_flow/static_quant.py:14]: should this still be from torchao.quantization.subclass?

**Review by jcaip — APPROVED** [review]

File: tutorials/calibration_flow/static_quant.py (hunk 1 above)
Comment (jerryzh168) [reply tutorials/calibration_flow/static_quant.py:14]: oh right, will update

**Comment by jerryzh168** [comment]
@jcaip please take a look again, the scope increased a bit to include a refactor for torch_function/torch_dispatch related utils functions

also @msaroufim please take a look as well

**Review by gau-nernst — APPROVED** [review]
The new `_dispatch__torch_function__` and `_dispatch__torch_dispatch__` are very nice, reduce boilerplate. Apart from also making the changes to `torchao/prototype/quant_llm/quant_llm.py`, for the codes that I wrote, as long as the tests pass, they are good for me.

**Review by msaroufim — APPROVED** [review]
nice clean change, just double checking this won't cause any BC issues on tune side for their nf4 recipe?

**Comment by jerryzh168** [comment]
> nice clean change, just double checking this won't cause any BC issues on tune side for their nf4 recipe?

oh we haven't changed nf4 yet, I can do it in a separate PR

**Pull request merged**